### PR TITLE
[CI] Only publish on release (tag) creation

### PR DESCRIPTION
**Pull request by Frenzie**
Cf. <https://github.com/hellerbarde/stapler/pull/67#discussion_r382946658>.

---
This change is [Reviewable](https://reviewable.io/reviews/hellerbarde/stapler/69)

### CODE REVIEW

File: .github/workflows/test-deploy.yml [hunk 1]
```diff
@@ -27,6 +32,6 @@ jobs:
       - name: Deploy to PyPI
         run: poetry publish
         continue-on-error: true
-        if: success() && github.event_name == 'push' && github.ref == 'refs/heads/master'
+        if: success() && github.event_name == 'release'
```
Comment (Frenzie) [inline .github/workflows/test-deploy.yml]: You could add further conditions, but unless there are tags you wish to whitelist or blacklist it comes down to the same thing. See https://help.github.com/en/actions/reference/contexts-and-expression-syntax-for-github-actions#functions
```suggestion
        if: success() && github.event_name == 'release' && startsWith(github.ref, 'refs/tags/')
```

**Comment by corollari** [comment]
I've tested this on my local fork and have been unable to trigger the publish step, I've tried:
- Push a tagged commit
- Create a github release

You can see the action runs in https://github.com/corollari/stapler/actions

**Comment by Frenzie** [comment]
@corollari You're not looking at the right run. ;-)

This one's "on: release".
https://github.com/corollari/stapler/runs/476324028?check_suite_focus=true

It doesn't trigger until you create a release (as you did); just the commit itself will do a "push" run.

**Comment by Frenzie** [comment]
Filter something like this: https://github.com/corollari/stapler/actions?query=workflow%3A%22Test+%26+Deploy%22+event%3Arelease

**Comment by corollari** [comment]
I see, you are right. Do you know why are there three different "on: push" runs tho?

**Comment by Frenzie** [comment]
I think it's commit push = push, tag push = push, tag fully processed = release.

But I'm not really sure why you'd have one more push than me. Force push? Same commit hash in a different branch?

You can see here they also have two pushes for the same reason, i.e. commit push & tag push (they don't build on release):
https://github.com/mifi/lossless-cut/runs/474592835

**Comment by Frenzie** [comment]
We could consider always running `echo ${{ github.ref }}` and the like as some kind of "environment" first step so you always know the details. (Tbh I'm a bit surprised you can't seem to coax this out of GH Actions anywhere.)

**Comment by hellerbarde** [comment]
this seems very clumsy. I tried to look into it as well, but couldn't find much useful either. Let's just "fix" it with our tooling, aka poetry just won't publish it until the release number is increased. It just means that we have to be very vigilant with updating the pyproject.toml file...

**Comment by Frenzie** [comment]
I'm not really sure which part you consider clumsy or why? It runs poetry publish when you tag a release without any workarounds.

**Comment by hellerbarde** [comment]
Oh. From the discussions you had with @corollari it sounded like it was a little tricky to trigger correctly. Apologies. I'll take another look.

**Comment by corollari** [comment]
Can confirm it works properly (note that it only publishes to pypi when a new GH release is made, not when a tag is pushed like in our previous discussion). My previous claims about it being hard to trigger were wrong, I got confused with github's UI.

**Comment by Frenzie** [comment]
Apologies, I didn't realize there was a difference between a tag and release. That's another GH UI issue right there: tags are prominently shown under releases, not presented as something distinct from releases. I thought a GH release was just a tag with optionally some additional metadata. Which is exactly what it is of course, just that one of those extra pieces of metadata says it's a release.

Judging by my reading of the GH Actions docs & GH event types ([here](https://developer.github.com/v3/activity/events/types)) that might mean a variation could be preferable instead, depending on whether you want to create "proper" GH releases or not. I.e., run on push all the same, but only publish if we're on a tag. Something like `github.event_name == 'push' && startsWith(github.ref, 'refs/tags/')`.

The rest of the discussion with @corollari is unrelated. I was just saying that the content of `$GITHUB_EVENT_PATH` were always printed (that's a file which contains the event JSON), we might have been able to ascertain what triggered that third push build. But except potentially to GitHub's wallet, a push build more or less is pretty harmless. :-)

**Comment by corollari** [comment]
I see, in my opinion using releases instead of tags also fulfills the original purpose (to a released version from not being in pypi) so this solution is also a good one.

**Comment by hellerbarde** [comment]
Yep, sounds good.

**Comment by Frenzie** [comment]
In that case the PR should be ready to go in its current state. :+1:

**Pull request merged**